### ISSUE TITLE
New progress bar

### ISSUE THREAD

**Issue by BoPeng** [vatlab/SoS#1059]
We had a broken progress bar for -v1 because things are running in parallel and because it is difficult to know the number of substeps in advance. With the new zmq-based controller, it is possible to do this:

assign a random character to each step (the first character of the step id, or sequentially), say 'a'
for each completed substep, output A for completed and a for skipped ..

**Comment by BoPeng** [comment]
The new progress bar simply output a `.` with the completion of each substep, green for regular and gray for skipped.

**Comment by gaow** [comment]
I'm on `master` now, and trying this out:

```
[GW] sos run wasp_to_fastqtl.sos -j8 -v1
................[GW] 
```

I like the coloring! However, it would be nice to have "hint phrase" at the beginning and a new line at the end? For the hint phrase, at least we will know how many substeps are completed, and we can report it like `[xx jobs completed] ...` Better avoid using `substeps` because it is a jargon less well known.

**Comment by BoPeng** [comment]
```
$ sos run test -v1 -s force
.............................. 2/20 completed, 1/10 ignored
$ sos run test -v1
.............................. 3/30 ignored
```

**Comment by gaow** [comment]
Looking good, but I'm confused by the numbers: shouldnt it be 20 completed, 10 ignored? I'm not sure what the fraction means here.

Here is my example running again:

```
[GW] sos run wasp_to_fastqtl.sos -v1
................ 1/16 ignored
[GW] sos run wasp_to_fastqtl.sos -v1 -s force
................[GW]
```

You see again I have 16 sub steps,all should be ignored. But I get a 1/16 number. Also if I run it again with `-s force` the output shows nothing ...

**Comment by BoPeng** [comment]
1 is the number of steps, 16 is substeps, we can ignore 16 but 1 does not match the number of dots. Not sure why the summary message is not sent with `-s force`.

**Comment by BoPeng** [comment]
For the second one, did the workflow execute successfully without `-v1`?

**Comment by gaow** [comment]
And the outcome is a bit redundant for nested workflow:

```
[preprocess]
output: "1.txt"
bash:
  touch 1.txt

[postprocess]
output: "2.txt"
bash:
  touch 2.txt

[default]
sos_run('preprocess+postprocess')
```

```
[GW] sos run test.sos -v1
.. 2/2 completed
 2/2 completed
```

The `2/2` completed showed up twice. It would be nice to only report it at the end, for once.

**Comment by BoPeng** [comment]
test again.

**Comment by BoPeng** [comment]
Tests are getting worse because of explicit terminating of threads and closing sockets, because whereas the `sos` command uses `Base_Executor()` once and release it, the tests use `Base_Executor()` multiple times, making resource management more difficult.

The problem with no response is partly due to python's interpret nature. When `sos` is started, it needs to reads tens of modules (small files) before doing anything, which can take a lot of time on systems with slow disk. This is a general problem with using python for CLI application. There are some tricks but generally not much can be done. sos also does not show anything during script and DAG parsing, which could be improved, I guess.

**Comment by gaow** [comment]
> Tests are getting worse because of explicit terminating of threads and closing sockets

Sorry to hear about it. I do get some hang issues on one of my workflow but maybe i'll wait for #1056 to be resolved before worrying about them. (should not bug you too soon).

> The problem with no response is partly due to python's interpret nature. 

No I believe it is SoS issue. Here is a MWE:

[wasp_to_fastqtl.sos.txt](https://github.com/vatlab/SoS/files/2407455/wasp_to_fastqtl.sos.txt)

if you run 

```
sos run wasp_to_fastqtl.sos -v1
```

you'll see what appears like a hang and perhaps can imaging the usefulness of a spinning rod.

**Comment by BoPeng** [comment]
This is because sos prints a dot **after** the completion of a substep, which in your example is 60s.

**Comment by gaow** [comment]
Exactly. That's all I'm complaining about (also applies to `-v2`)

**Comment by gaow** [comment]
>  I do get some hang issues on one of my workflow but maybe i'll wait for #1056 to be resolved before worrying about them. (should not bug you too soon).

This is a real hang, though, but I'll try to make MWE after the branch is finalized -- hopefully wont be an issue by then!

**Comment by BoPeng** [comment]
So you want something like a rotating line during the execution and replace it with a dot after the completion?

**Comment by gaow** [comment]
Yes, I think that'd be a way to make it less confusing and I'm surely open to better suggestions!

**Comment by BoPeng** [comment]
rotating on terminal is difficult, essentially needing a thread that outputs `\b` and some sequence of characters... and will make a huge mess if the output is piped to a file.. There is  [an example  code](https://stackoverflow.com/questions/4995733/how-to-create-a-spinning-command-line-cursor-using-python) but I do not know if I want to do it.

**Comment by gaow** [comment]
Then a blinking color block might help, or even a blinking dot? Although a blinking dot would look weird in `-v2`

**Comment by BoPeng** [comment]
blinking ansi code does not work on my terminal...

```
$ echo -e "\033[5mHello\033[0m"
Hello
$ echo -e "\033[32mHello\033[0m"
Hello
```
The second one changes `Hello` to green.

**Comment by gaow** [comment]
Neither do they work on my end ... then maybe 3 green dots upfront `...` is better than nothing. And they will be replaced by whatever first real dot comes in ('\r').

**Comment by BoPeng** [comment]
```
$ sos run test -v1 -s force
=====>
```

**Comment by gaow** [comment]
I got a non-blinking single green box now! I think it is good enough (though still feel like 3 is nicer ...)

I observe that the green box will show, followed by a blinking cursor, which blinks for about 10 sec, then become solid (like hanging) for the next 50 secs. Not sure if you see the same on your computer. But this is a trivial point compared to more interesting problems in #1062

**Comment by BoPeng** [comment]
The brighter box means "working" and turns to darker box for completed. I however have a problem with progressbar not wrapping to next line so it will stuck at the end of the line if there are many substeps.

**Comment by gaow** [comment]
For the displayed messages, I understand the "structural" argument and I agree. But I think some re-phrasing will make it shorter and clearer:

```
20 steps processed (220 jobs completed)
20 steps processed (20 jobs completed, 150 jobs ignored)
```

that avoids the "from" statement and makes it shorter. As you said, jobs are consequences of steps, whether it be local runs or remote tasks. 

Also I'm rerunning one of my analysis pipeline and realized it might be better to print blocks for steps on the screen, now that we have reported substeps/jobs. Because my output contains >30k units (transcripts), the screen is a disaster ...

**Comment by BoPeng** [comment]
We have a typical unknown total size problem. Changing to step is possible but you would not see any update if you have a step with a large number of substeps. Perhaps we can show alternating symbols (`\b`) for substeps and progress the progress bar with the completion of steps. It is also needed to control the alternating of symbols (e.g. at most 1 per second) to avoid heavy traffic to terminal.

**Comment by gaow** [comment]
Sounds like good idea. Yes, too much traffic on the terminal hurts, considering the 10K job situation, many of them will end approximately the same time but have to wait until all blocks are written on screen. It's not optimal.

**Comment by BoPeng** [comment]
`-v2` is not an issue, right? Previously, only `_output` will be displayed after the completion of steps but right now sos will print something like "executing ... step".

**Comment by gaow** [comment]
> Previously, only _output will be displayed after the completion of steps but right now sos will print something like "executing ... step".

Yes that would be sufficient for `-v2`.  The new progress bar behavior looks great!
